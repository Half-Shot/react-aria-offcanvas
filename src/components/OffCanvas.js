import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as focusManager from '../helpers/focusManager';
import createStyles from '../helpers/styles';

const ESC_KEY = 27;

export default class OffCanvas extends Component {
  static defaultProps = {
    isOpen: false,
    width: '300px',
    height: '300px',
    position: 'left',
    closeOnEsc: true,
    closeOnOverlayClick: true,
    trapFocusAfterOpen: true,
    returnFocusAfterClose: true,
    style: {
      overlay: {},
      content: {},
    },
  };

  static propTypes = {
    isOpen: PropTypes.bool.isRequired,
    width: PropTypes.string,
    height: PropTypes.string,
    position: PropTypes.oneOf(['left', 'right', 'top', 'bottom']),
    onClose: PropTypes.func,
    closeOnEsc: PropTypes.bool,
    closeOnOverlayClick: PropTypes.bool,
    trapFocusAfterOpen: PropTypes.bool,
    returnFocusAfterClose: PropTypes.bool,
    style: PropTypes.shape({
      overlay: PropTypes.object,
      content: PropTypes.object,
    }),
    className: PropTypes.string,
    overlayClassName: PropTypes.string,
    role: PropTypes.string,
    label: PropTypes.string,
    labelledby: PropTypes.string,
    children: PropTypes.node,
  };

  static defaultStyles = {
    overlay: {
      position: 'fixed',
      top: 0,
      left: 0,
      width: '100%',
      height: '100%',
      zIndex: '900',
    },
    content: {
      position: 'fixed',
      zIndex: '1000',
      outline: 0,
    },
  };

  static extraStyles = {
    overlay: {
      background: 'rgba(255, 255, 255, 0.5)',
    },
    content: {
      background: 'rgba(0, 0, 0, 0.1)',
      transition: '0.25s ease-out',
    },
  };

  componentDidMount() {
    if (this.props.isOpen) {
      this.open();
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props.isOpen && !prevProps.isOpen) {
      this.open();
    } else if (!this.props.isOpen && prevProps.isOpen) {
      this.close();
    }
  }

  open = () => {
    if (this.props.returnFocusAfterClose) {
      focusManager.focusLater();
    }

    if (this.props.trapFocusAfterOpen) {
      focusManager.setFocusTrap(this.content);
    }

    this.focusContent();
  };

  close = () => {
    if (this.props.returnFocusAfterClose) {
      focusManager.returnFocus();
    }

    if (this.props.trapFocusAfterOpen) {
      focusManager.removeFocusTrap();
    }
  };

  parentHandlesClose = event => {
    if (this.props.onClose) {
      this.props.onClose(event);
    }
  };

  handleOnOverlayClick = event => {
    if (this.props.closeOnOverlayClick && event.target === this.overlay) {
      this.parentHandlesClose(event);
    }
  };

  handleOnEscDown = event => {
    if (this.props.closeOnEsc && event.keyCode === ESC_KEY) {
      this.parentHandlesClose(event);
    }
  };

  setOverlayRef = overlay => {
    this.overlay = overlay;
  };

  setContentRef = content => {
    this.content = content;
  };

  focusContent = () => this.content && this.content.focus();

  getExtraStyles = () => {
    const { className, overlayClassName } = this.props;

    // Remove extra styles when classNames are passed
    const overlayStyles = overlayClassName ? {} : OffCanvas.extraStyles.overlay;
    const contentStyles = className ? {} : OffCanvas.extraStyles.content;

    return {
      overlay: overlayStyles,
      content: contentStyles,
    };
  };

  getStyles = () => {
    const { isOpen, width, height, position, style } = this.props;
    const extraStyles = this.getExtraStyles();

    const styles = createStyles(
      OffCanvas.defaultStyles,
      extraStyles,
      isOpen,
      width,
      height,
      position,
      style,
    );

    return styles;
  };

  render() {
    const {
      isOpen,
      role,
      label,
      labelledby,
      className,
      overlayClassName,
    } = this.props;

    const styles = this.getStyles();

    return (
      <div
        ref={this.setOverlayRef}
        style={styles.overlay}
        className={overlayClassName}
        onClick={this.handleOnOverlayClick}
        data-testid="overlay"
      >
        <div
          ref={this.setContentRef}
          style={styles.content}
          className={className}
          onKeyDown={this.handleOnEscDown}
          role={role}
          aria-label={label}
          aria-labelledby={labelledby}
          aria-hidden={!isOpen}
          tabIndex="-1"
          data-testid="content"
        >
          {this.props.children}
        </div>
      </div>
    );
  }
}
