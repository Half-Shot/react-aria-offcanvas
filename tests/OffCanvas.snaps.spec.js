import React from 'react';
import renderer from 'react-test-renderer';
import OffCanvas from 'react-off-canvas';

describe('OffCanvas', () => {
  it('renders correctly when open', () => {
    const tree = renderer.create(<OffCanvas isOpen />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when closed', () => {
    const tree = renderer.create(<OffCanvas />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
