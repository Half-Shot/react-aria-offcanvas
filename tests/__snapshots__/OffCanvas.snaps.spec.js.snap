// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OffCanvas renders correctly when closed 1`] = `
<div
  className={undefined}
  data-testid="overlay"
  onClick={[Function]}
  style={
    Object {
      "background": "rgba(255, 255, 255, 0.5)",
      "height": "100%",
      "left": 0,
      "position": "fixed",
      "top": 0,
      "width": "",
      "zIndex": "900",
    }
  }
>
  <div
    aria-hidden={true}
    aria-label={undefined}
    aria-labelledby={undefined}
    className={undefined}
    data-testid="content"
    onKeyDown={[Function]}
    role={undefined}
    style={
      Object {
        "background": "rgba(0, 0, 0, 0.1)",
        "height": "300px",
        "left": "0",
        "outline": 0,
        "position": "fixed",
        "transform": "translateX(-100%)",
        "transition": "transform 0.25s ease-out",
        "transitionProperty": "",
        "width": "300px",
        "zIndex": "1000",
      }
    }
    tabIndex="-1"
  />
</div>
`;

exports[`OffCanvas renders correctly when open 1`] = `
<div
  className={undefined}
  data-testid="overlay"
  onClick={[Function]}
  style={
    Object {
      "background": "rgba(255, 255, 255, 0.5)",
      "height": "100%",
      "left": 0,
      "position": "fixed",
      "top": 0,
      "width": "100%",
      "zIndex": "900",
    }
  }
>
  <div
    aria-hidden={false}
    aria-label={undefined}
    aria-labelledby={undefined}
    className={undefined}
    data-testid="content"
    onKeyDown={[Function]}
    role={undefined}
    style={
      Object {
        "background": "rgba(0, 0, 0, 0.1)",
        "height": "300px",
        "left": "0",
        "outline": 0,
        "position": "fixed",
        "transform": "",
        "transition": "transform 0.25s ease-out",
        "transitionProperty": "transform",
        "width": "300px",
        "zIndex": "1000",
      }
    }
    tabIndex="-1"
  />
</div>
`;
